src/adapter/src/coord.rs

```diff
@@ -1635,12 +1578,191 @@ impl Coordinator {
+                    catalog.set_optimized_plan(id, optimized_plan);
+                    catalog.set_physical_plan(id, physical_plan);
+                    catalog.set_dataflow_metainfo(id, metainfo);
+
+                    compute_instance.insert_collection(id);
+                }
+                CatalogItem::MaterializedView(mv) => {
+                    // Collect optimizer parameters.
+                    let compute_instance =
+                        instance_snapshots.entry(mv.cluster_id).or_insert_with(|| {
+                            self.instance_snapshot(mv.cluster_id)
+                                .expect("compute instance exists")
+                        });
+                    let internal_view_id = self.allocate_transient_id()?;
+                    let debug_name = self
+                        .catalog()
+                        .resolve_full_name(entry.name(), None)
+                        .to_string();
+
+                    // Build an optimizer for this MATERIALIZED VIEW.
+                    let mut optimizer = optimize::materialized_view::Optimizer::new(
+                        self.owned_catalog(),
+                        compute_instance.clone(),
+                        entry.id(),
+                        internal_view_id,
+                        mv.desc.iter_names().cloned().collect(),
+                        mv.non_null_assertions.clone(),
+                        debug_name,
+                        optimizer_config.clone(),
+                    );
+
+                    // MIR ⇒ MIR optimization (global)
+                    let global_mir_plan = optimizer.optimize(mv.optimized_expr.clone())?;
+                    let optimized_plan = global_mir_plan.df_desc().clone();
+
+                    // MIR ⇒ LIR lowering and LIR ⇒ LIR optimization (global)
+                    let global_lir_plan = optimizer.optimize(global_mir_plan)?;
+                    let physical_plan = global_lir_plan.df_desc().clone();
+                    let metainfo = global_lir_plan.df_meta().clone();
+
+                    let catalog = self.catalog_mut();
+                    catalog.set_optimized_plan(id, optimized_plan);
+                    catalog.set_physical_plan(id, physical_plan);
+                    catalog.set_dataflow_metainfo(id, metainfo);
+
+                    compute_instance.insert_collection(id);
+                }
+                _ => (),
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Collects for each index the materialized views that depend on it, either directly or
+    /// transitively through other indexes (but not through other MVs).
+    ///
+    /// The returned information is required during coordinator bootstrap for index as-of
+    /// selection, to ensure that selected as-ofs satisfy the requirements of downstream MVs.
+    ///
+    /// This method expects all dataflow plans to be available, so it must run after
+    /// `bootstrap_dataflow_plans`.
```

Comment:
I think you can use something like [`Coordinator::bootstrap_dataflow_plans()`] to make sure that this reference doesn't rot.